// local/repositories/leaderboard/leaderboard-segment.js
'use strict';

var moment = require('moment');

var LeaderboardSegment = function (id, data) {
    if (!data) data = {};

    this.id = id;

    this.periodStart = data.ps ? data.ps : 0;
    this.periodEnd = data.pe ? data.pe : 0;

    this.wormholesAdded = data.wa ? data.wa : 0;
    this.wormholesPartCompleted = data.wpc ? data.wpc : 0;
    this.wormholesCompleted = data.wc ? data.wc : 0;

    this.foreignWormholesPartCompleted = data.fwpc ? data.fwpc : 0;
    this.foreignWormholesCompleted = data.fwc ? data.fwc : 0;

    this.systemsUpdated = data.su ? data.su : 0;
    this.systemsPartScanned = data.sps ? data.sps : 0;
    this.systemsFullScanned = data.sfs ? data.sfs : 0;

    this.likes = data.l ? data.l : 0;

    Object.defineProperties(this, {
        totalWormholesPartCompleted: {
            get: function () {
                return this.wormholesPartCompleted
                  + this.foreignWormholesPartCompleted;
            }
        },
        totalWormholesCompleted: {
            get: function () { return this.wormholesCompleted + this.foreignWormholesCompleted; }
        }
    });

}
LeaderboardSegment.prototype.toRemoteForm = function () {
    return {
        ps: this.periodStart,
        pe: this.periodEnd,
        wa: this.wormholesAdded,
        wpc: this.wormholesPartCompleted,
        wc: this.wormholesCompleted,
        fwpc: this.foreignWormholesPartCompleted,
        fwc: this.foreignWormholesCompleted,
        su: this.systemsUpdated,
        sps: this.systemsPartScanned,
        sfs: this.systemsFullScanned,
        l: this.likes
    };
}
LeaderboardSegment.prototype.timeInPeriod = function (time) {
	return time >= this.periodStart && time < this.periodEnd;
}
LeaderboardSegment.prototype.add = function (other) {

    if (other.wormholesAdded) this.wormholesAdded += other.wormholesAdded;
    if (other.wormholesPartCompleted) this.wormholesPartCompleted += other.wormholesPartCompleted;
    if (other.wormholesCompleted) this.wormholesCompleted += other.wormholesCompleted;
    
    if (other.foreignWormholesPartCompleted)
        this.foreignWormholesPartCompleted += other.foreignWormholesPartCompleted;
    if (other.foreignWormholesCompleted)
        this.foreignWormholesCompleted += other.foreignWormholesCompleted;

    if (other.systemsUpdated) this.systemsUpdated += other.systemsUpdated;
    if (other.systemsPartScanned) this.systemsPartScanned += other.systemsPartScanned;
    if (other.systemsFullScanned) this.systemsFullScanned += other.systemsFullScanned;
    
    if (other.likes) this.likes += other.likes;

    return this;
}

LeaderboardSegment.prototype.getPeriodStartPretty = function () {
    return moment(this.periodStart).zone('+0000').format('Do MMM, dddd');
}

module.exports = LeaderboardSegment;
