// local/repositories/leaderboard/leaderboard-repository.js
'use strict';

var baseRepository = require('../base');
var Leaderboard = require('./leaderboard');

var base = baseRepository({
    path: 'leaderboard-new',
    makeLocal: function (id, data) {
        return new Leaderboard(id, data);
    },
    makeRemote: function (local) {
        var days = {};
        for (var key in local.days) {
            days[key] = local.days[key].toRemoteForm();
        }

        var json = {
            allTime: local.allTime ? local.allTime.toRemoteForm() : null,
            findsPaidFor: local.findsPaidFor ? local.findsPaidFor.toRemoteForm() : null,
            totalPaidAmount: local.totalPaidAmount,
            days: days,
            lastModified: Date.now()
        }
        return json;
    },
    pruneRemote: true
});

var me = {
    //get: base.get,
    getAll: base.getAll,
    getArray: base.getArray,
    //create: base.create,
    update: base.update,
    'delete': base.delete,
    changed: base.changed,
    changedRemoveHandler: base.changedRemoveHandler,
    initialize: base.initialize,
    supressChanged: base.supressChanged
};
me.get = function (id) {
    var res = base.get(id);

    if (!res) res = new Leaderboard(id);
    return res;
}

module.exports = me;
