// local/repositories/wormhole-import-record/wormhole-import-record-repository.js
'use strict';

var baseRepository = require('../base');

var Log = function (id, data) {
    this.id = id;
    this.type = data.type;
    this.lastModified = data.mod;
    this.foreignId = data.fId;
}

var base = baseRepository({
    path: 'wormhole-import-records',
    makeLocal: function (id, data) {
        return new Log(id, data);
    },
    makeRemote: function (local) {
        var json = {
            type: local.type,
            mod: new Date().getTime(),
            fId: local.foreignId
        }
        return json;
    }
});

var me = {
    get: base.get,
    getAll: base.getAll,
    getArray: base.getArray,
    //create: base.create,
    //update: base.update,
    'delete': base.delete,
    changed: base.changed,
    changedRemoveHandler: base.changedRemoveHandler,
    initialize: base.initialize,
    supressChanged: base.supressChanged
};
me.set = function (type, foreignId) {
    var itm = new Log(null, {
        type: type,
        fId: foreignId
    });
    base.update(itm);
}

module.exports = me;
