// local/repositories/login-log/login-log-repository.js
'use strict';

var baseRepository = require('../base');

var LoginLog = function (id, data) {
    this.id = id ? id : (new Date()).getTime();
    this.charId = data.charId;
    this.charName = data.charName;
    this.initCharId = data.initCharId;
    this.initCharName = data.initCharName;
}

var base = baseRepository({
    path: 'login-log',
    makeLocal: function (id, data) {
        return new LoginLog(id, data);
    },
});

var me = {
    get: base.get,
    getAll: base.getAll,
    getArray: base.getArray,
    //create: base.create,
    //update: base.update,
    'delete': base.delete,
    changed: base.changed,
    changedRemoveHandler: base.changedRemoveHandler,
    initialize: base.initialize,
    supressChanged: base.supressChanged
};
me.set = function (charId, charName,initCharId,initCharName) {
    var itm = new LoginLog(null, {
        charId: charId,
        charName: charName,
        initCharId: initCharId,
        initCharName: initCharName
    });
    base.update(itm);
}

module.exports = me;
