// local/repositories/authorization/authorization-repository.js
'use strict';

var baseRepository = require('../base');

var Authorization = require('./authorization.js');

var base = baseRepository({
    path: 'authorizations',
    makeLocal: function (id, data) {
        return new Authorization(id, data);
    },
    makeRemote: function (local) {
        var json = {
            name: local.name,
            type: local.type,
            mod: Date.now(),
            note: local.note
        }
        return json;
    },
    pruneRemote: true
});

var me = {
    get: base.get,
    getAll: base.getAll,
    getArray: base.getArray,
    //create: base.create,
    //update: base.update,
    'delete': base.delete,
    changed: base.changed,
    changedRemoveHandler: base.changedRemoveHandler,
    initialize: base.initialize,
    supressChanged: base.supressChanged
};
me.set = function (id,name,type,note) {
    var itm = new Authorization(id, {
        name: name,
        type: type,
        note: note
    });
    base.update(itm);
}

module.exports = me;
